e2e/nomostest/reset.go

```diff
@@ -0,0 +1,531 @@
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"kpt.dev/configsync/pkg/api/configmanagement"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/importer/filesystem"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/metrics"
+	"kpt.dev/configsync/pkg/reconcilermanager"
+	"kpt.dev/configsync/pkg/syncer/differ"
+	"kpt.dev/configsync/pkg/util/log"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Reset performs multi-repo test reset:
+// - Delete unmanaged RootSyncs & RepoSyncs (with deletion propagation)
+// - Validate managed RepoSyncs & RootSyncs are deleted
+// - Delete all test namespaces not containing config-sync itself
+// - Delete NonRootRepos & RootRepos (except the default RootRepo)
+//
+// This should cleanly delete or reset all registered RSyncs.
+// Any managed RSyncs must have deletion propagation enabled by the test that
+// created them, otherwise their managed resources will not be deleted when the
+// RSync is deleted.
+// Any unregistered Repos must be reset by individual test Cleanup.
+func Reset(nt *NT) error {
+	start := time.Now()
+	defer func() {
+		elapsed := time.Since(start)
+		nt.T.Logf("[RESET] Test environment reset took %v", elapsed)
+	}()
+
+	// Delete all existing RootSyncs with the test label.
+	// Enable deletion propagation first, to clean up managed resources.
+	rootSyncList, err := listRootSyncs(nt)
+	if err != nil {
+		return err
+	}
+	if err := ResetRootSyncs(nt, rootSyncList.Items); err != nil {
+		return err
+	}
+
+	// Delete all existing RepoSyncs with the test label.
+	// Enable deletion propagation first, to clean up managed resources.
+	repoSyncList, err := listRepoSyncs(nt)
+	if err != nil {
+		return err
+	}
+	if err := ResetRepoSyncs(nt, repoSyncList.Items); err != nil {
+		return err
+	}
+
+	// Delete all Namespaces with the test label (except shared).
+	nsList, err := listNamespaces(nt)
+	if err != nil {
+		return err
+	}
+	// Error if any protected namespace was modified by a test (test label added)
+	// and not reverted by the test.
+	protectedNamespaces, nsListItems := filterNamespaces(nsList.Items,
+		protectedNamespaceList()...)
+	if len(protectedNamespaces) > 0 {
+		return errors.Errorf("protected namespace(s) modified by test: %+v",
+			protectedNamespaces)
+	}
+	// Skip deleting namespaces that contain test infra or Config Sync itself.
+	_, nsListItems = filterNamespaces(nsListItems,
+		configsync.ControllerNamespace,
+		configmanagement.RGControllerNamespace,
+		metrics.MonitoringNamespace,
+		testGitNamespace)
+	if err := ResetNamespaces(nt, nsListItems); err != nil {
+		return err
+	}
+
+	// NOTE: These git repos are not actually being deleted here, just forgotten.
```

Comment:
What is the reason that we need to forget the git repos here? It seems a bit strange to me that these pointers are being removed without cleaning up the underlying git repo.

Minor nit on the second line of the comment: it looks like the repos should normally get deleted by [this cleanup block](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/new.go#L319). The logic in `Clean` seems to be for stale repos (older than 24 hours)

Reply:
Repos were never actually being **deleted** before either:
https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/new.go#L234-L239
https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/config_sync.go#L1105
https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/config_sync.go#L1131

It WAS "resetting" the repo's contents (partially) and syncing again, but that never worked well (only some known files were being removed) and it's no longer necessary because the new finalizer does a destroy based on what's in the inventory, not what's in the repo, and then the repo contents are reset in test setup with ReInit:
https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/config_sync.go#L1062

This new way, we skip resetting repos we don't plan to use in the next test.